lib/id_partner.rb

```diff
@@ -0,0 +1,278 @@
+      }
+
+      # Send the userinfo request
+      userinfo_request = Net::HTTP::Get.new(uri.request_uri, headers)
+      userinfo_response = http.request(userinfo_request)
+
+      # Check the response and parse the userinfo data
+      raise "Failed to retrieve userinfo: #{userinfo_response.body}" unless userinfo_response.is_a?(Net::HTTPSuccess)
+
+      JSON.parse(userinfo_response.body)
+    end
+
+    private
+
+    def extract_claims(claims_object)
+      return [] unless claims_object.is_a?(Hash)
+
+      userinfo_keys = claims_object[:userinfo]&.keys || []
+      id_token_keys = claims_object[:id_token]&.keys || []
+      (userinfo_keys + id_token_keys).uniq
+    end
+
+    def obtain_well_known_config_endpoints
+      uri = URI("#{config[:provider_url]}/.well-known/openid-configuration")
+      response = Net::HTTP.get(uri)
+      well_known_config = JSON.parse(response)
+      @endpoints = {
+        authorization_endpoint: well_known_config["authorization_endpoint"],
+        token_endpoint: well_known_config["token_endpoint"],
+        userinfo_endpoint: well_known_config["userinfo_endpoint"],
+        pushed_authorization_request_endpoint: well_known_config["pushed_authorization_request_endpoint"],
+        jwks_uri: well_known_config["jwks_uri"]
+      }
+    rescue StandardError => e
+      raise "Failed to fetch well-known config: #{e.message}"
+    end
+
+    def create_request_object(params)
+      extended_params = params.merge({
+                                       iss: config[:client_id],
+                                       aud: config[:provider_url],
+                                       exp: Time.now.to_i + 60,
+                                       iat: Time.now.to_i,
+                                       nbf: Time.now.to_i
+                                     })
+
+      sig_key = sig_key()
+      jwt = JSON::JWT.new(extended_params)
+      jwt.sign(sig_key, sig_key["alg"]).to_s
+    end
+
+    def push_authorization_request(request_object)
+      uri = URI(endpoints[:pushed_authorization_request_endpoint])
+      http = Net::HTTP.new(uri.host, uri.port)
+      basic_auth_credentials = Base64.strict_encode64("#{config[:client_id]}:#{config[:client_secret]}")
+      headers = {
+        "Authorization" => "Basic #{basic_auth_credentials}",
+        "Content-Type" => "application/x-www-form-urlencoded",
+        "Accept" => "application/json"
+      }
+      par_request = Net::HTTP::Post.new(uri.request_uri, headers)
+      par_request.set_form_data(request_object)
+      par_response = http.request(par_request)
+      raise "Failed to push authorization request: #{par_response.body}" unless par_response.is_a?(Net::HTTPSuccess)
+
+      JSON.parse(par_response.body)
+    rescue StandardError => e
+      raise "An error occurred: #{e.message}"
+    end
+
+    def generate_code_challenge(code_verifier)
+      Base64.urlsafe_encode64(Digest::SHA256.digest(code_verifier)).gsub(/=+$/, "")
```

Comment:
similar to node: https://github.com/panva/node-openid-client/blob/a84d022f195f82ca1c97f8f6b2567ebcef8738c3/lib/helpers/generators.js#L13C22-L13C39